odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1471 @@
+        ----------
+        func : `Functional`
+            Functional to which the scalar is added.
+        scalar : `element` in the `field` of the ``domain``
+            The scalar to be added to the functional. The `field` of the
+            ``domain`` is the range of the functional.
+        tmp_dom : `LinearSpaceVector`, optional
+            ...
+        """
+        # TODO: what is tmp_dom used for?
+        if not isinstance(func, Functional):
+            raise TypeError('functional {!r} is no a Functional instance'
+                            ''.format(func))
+        if scalar not in func.range:
+            raise TypeError('the scalar {} is not in the range of the'
+                            'functional {!r}'.format(scalar, func))
+
+        Functional.__init__(self, domain=func.domain, linear=func.is_linear,
+                            smooth=func.is_smooth, concave=func.is_concave,
+                            convex=func.is_convex,
+                            grad_lipschitz=func.grad_lipschitz)
+
+        OperatorSum.__init__(self, left=func,
+                             right=ConstantOperator(vector=scalar,
+                                                    domain=func.domain,
+                                                    range=func.range),
+                             tmp_ran=None, tmp_dom=tmp_dom)
+
+        self.scalar = scalar
+
+    @property
+    def gradient(self):
+        """Gradient operator of the FunctionalScalarSum.
+
+        This is the same as the gradient of the original functional.
+        """
+        return self.left.gradient
+
+    def proximal(self, sigma=1.0):
+        """Proximal operator of the FunctionalScalarSum.
+
+        This is the same as the proximal operator of the original
+        functional.
+
+        Parameters
+        ----------
+        sigma : positive float, optional
+            Regularization parameter of the proximal operator.
+
+        Returns
+        -------
+        out : `Operator`
+            Domain and range equal to domain of functional.
+        """
+        return self.left.proximal(sigma)
+
+    @property
+    def conjugate_functional(self):
+        """Convex conjugate functional of FunctionalScalarSum."""
+        return self.left.conjugate_functional - self.scalar
+
+    def derivative(self, point):
+        """Returns the derivative operator of FunctionalScalarSum.
+
+        Parameters
+        ----------
+        point : `LinearSpaceVector`
+            The point in which the gradient is evaluated.
+
+        Returns
+        -------
+        out : `DerivativeOperator`
+            The linear operator that maps ``x --> <x, grad_f(point)>``.
+        """
+        return self.left.derivative(point)
+
+
+class TranslatedFunctional(Functional):
+    """Implementation of the translated functional.
+
+    Given a functional ``f`` and an element ``translation`` in the domain of
+    ``f``, this corresponds to the functional ``f(. - translation)``.
```

Comment:
I had a short discussion with @kohr-h on which convention to use (so not the naming), and we concluded that this was probably the most natural one. Moreover the "translation helper functions" for proximals and convex conjugates also use the same convention.

But I agree: the "direction of translation" is non-obvious from the naming. If we have a better naming this should be applied in the other places too :)